pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -496,7 +502,11 @@ public boolean index(GenericRow row, @Nullable RowMetadata rowMetadata)
       // segment indexing or addNewRow call errors out in those scenario, there can be metadata inconsistency where
       // a key is pointing to some other key's docID
       // TODO fix this metadata mismatch scenario
-      if (_partitionUpsertMetadataManager.addRecord(this, recordInfo)) {
+      boolean isOutOfOrderRecord = _partitionUpsertMetadataManager.addRecord(this, recordInfo);
+      if (_upsertOutOfOrderRecordColumn != null) {
```

Comment:
(MAJOR) We need to check `isOutOfOrderRecord` first. Please add a test for this new feature

Reply:
Umm why? This if clause is for populating `_upsertOutOfOrderRecordColumn` field specifically. 
Not adding record check is at line 509.

Reply:
Added UTs for this.

Reply:
Sorry I misread it. I thought we should only set the value when it is out of order